A bad jsonpatch can knock out the CAPI controller manager

**What steps did you take and what happened:**
[A clear and concise description on how to REPRODUCE the bug.]
Add a list variable to the ClusterClass
```
  - name: controlPlanePreKubeadmCommands
    required: false
    schema:
      openAPIV3Schema:
        type: array
        items:
          type: string
        default: []
```
Add the following inline patch 
```
  - name: controlPlanePreKubeadmCommands
    definitions:
    - selector:
        apiVersion: controlplane.cluster.x-k8s.io/v1beta1
        kind: KubeadmControlPlaneTemplate
        matchResources:
          controlPlane: true
      jsonPatches:
      - op: replace
        path: /spec/template/spec/kubeadmConfigSpec/preKubeadmCommands
        valueFrom:
          template: |
            {{- range .controlPlanePreKubeadmCommands }}
            - {{ . }}
            {{- end }}
```
Omit controlPlanePreKubeadmCommands in your Cluster variables and create the Cluster. You will see CAPI controller manager Pod is down.
```
❯ kubectl get po -n capi-system
NAME                                      READY   STATUS             RESTARTS        AGE
capi-controller-manager-8778fdfcb-g57fk   0/1     CrashLoopBackOff   7 (2m21s ago)   16m
```
It crashes for this error. You can delete CAPI controller manager Pod for recreating but new Pod always hit this error.
``` 
I0206 02:15:26.100037       1 controller.go:117] "Observed a panic in reconciler: runtime error: invalid memory address or nil pointer dereference" controller="topology/cluster" controllerGroup="cluster.x-k8s.io" controllerKind="Cluster" cluster="default/workload-1" namespace="default" name="workload-1" reconcileID=e1cf4673-ef91-412e-94e5-42467d62e045
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1659c95]

goroutine 1328 [running]:
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()
        sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:118 +0x1f4
panic({0x1c367c0, 0x322b990})
        runtime/panic.go:838 +0x207
github.com/evanphx/json-patch/v5.findObject(0x778d6d?, {0xc000a33a00?, 0x1aeab00?}, 0xc0011901a0?)
        github.com/evanphx/json-patch/v5@v5.6.0/patch.go:513 +0x155
github.com/evanphx/json-patch/v5.Patch.add({0xc0003e6800?, 0x16b6?, 0x16b6?}, 0xc001aff768?, 0xc0019d4880?, 0xc0015c8be8)
        github.com/evanphx/json-patch/v5@v5.6.0/patch.go:714 +0xae
github.com/evanphx/json-patch/v5.Patch.ApplyIndentWithOptions({0xc000747d10?, 0x6, 0x6}, {0xc0003e6800, 0x16b6, 0x16b6}, {0x0, 0x0}, 0xc00094fec0?)
        github.com/evanphx/json-patch/v5@v5.6.0/patch.go:1095 +0x1f9
github.com/evanphx/json-patch/v5.Patch.ApplyWithOptions(...)
        github.com/evanphx/json-patch/v5@v5.6.0/patch.go:1059
github.com/evanphx/json-patch/v5.Patch.Apply({0xc000747d10, 0x6, 0x6}, {0xc0003e6800, 0x16b6, 0x16b6})
        github.com/evanphx/json-patch/v5@v5.6.0/patch.go:1053 +0x8e
sigs.k8s.io/cluster-api/internal/controllers/topology/cluster/patches.applyPatchesToRequest({0x21fd060?, 0xc00107dc80?}, 0xc0010a6500, 0xc00152c180)
        sigs.k8s.io/cluster-api/internal/controllers/topology/cluster/patches/engine.go:310 +0x32e
sigs.k8s.io/cluster-api/internal/controllers/topology/cluster/patches.(*engine).Apply(0xc0003babc0, {0x21fd060, 0xc0016c4420}, 0xc0018a8120, 0xc0011665c0)
        sigs.k8s.io/cluster-api/internal/controllers/topology/cluster/patches/engine.go:101 +0x39b
sigs.k8s.io/cluster-api/internal/controllers/topology/cluster.(*Reconciler).computeDesiredState(0xc00011cb00, {0x21fd060, 0xc0016c4420}, 0xc0018a80c0)
        sigs.k8s.io/cluster-api/internal/controllers/topology/cluster/desired_state.go:103 +0x431
sigs.k8s.io/cluster-api/internal/controllers/topology/cluster.(*Reconciler).reconcile(0xc000a8bdc0?, {0x21fd060, 0xc0016c4420}, 0xc0018a80c0)
        sigs.k8s.io/cluster-api/internal/controllers/topology/cluster/cluster_controller.go:235 +0x14a
sigs.k8s.io/cluster-api/internal/controllers/topology/cluster.(*Reconciler).Reconcile(0xc00011cb00, {0x21fd060, 0xc0016c4420}, {{{0xc000c042e0?, 0x10?}, {0xc000c042d6?, 0x40d987?}}})
        sigs.k8s.io/cluster-api/internal/controllers/topology/cluster/cluster_controller.go:198 +0x445
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0x21fcfb8?, {0x21fd060?, 0xc0016c4420?}, {{{0xc000c042e0?, 0x1de9760?}, {0xc000c042d6?, 0x4041f4?}}})
        sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:121 +0xc8
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0005d85a0, {0x21fcfb8, 0xc000991b00}, {0x1cc51e0?, 0xc0001c1820?})
        sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:320 +0x33c
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0005d85a0, {0x21fcfb8, 0xc000991b00})
        sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:273 +0x1d9
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
        sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:234 +0x85
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2
        sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:230 +0x325
```
You can not delete the problematic cluster object either since validating webhook is also in the crashed pod.
```
❯ kubectl delete cluster workload-1
Error from server (InternalError): Internal error occurred: failed calling webhook "validation.cluster.cluster.x-k8s.io": failed to call webhook: Post "https://capi-webhook-service.capi-system.svc:443/validate-cluster-x-k8s-io-v1beta1-cluster?timeout=10s": dial tcp 100.68.35.5:443: connect: connection refused
```
There is no way to recover CAPI controller manager unless you delete CAPI's  validatingwebhookconfiguration first.

**What did you expect to happen:**
CAPI controller manager Pod can tolerate errors from computing topology. Or provide a guidance for recovery.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]


**Environment:**

- Cluster-api version: v1.2.8
- minikube/kind version:
- Kubernetes version: (use `kubectl version`):
- OS (e.g. from `/etc/os-release`):

/kind bug
[One or more /area label. See https://github.com/kubernetes-sigs/cluster-api/labels?q=area for the list of labels]

--- a/internal/controllers/topology/cluster/patches/engine.go
+++ b/internal/controllers/topology/cluster/patches/engine.go
@@ -19,10 +19,13 @@ package patches
 
 import (
 	"context"
+	"fmt"
+	"runtime/debug"
 	"strings"
 
 	jsonpatch "github.com/evanphx/json-patch/v5"
 	"github.com/pkg/errors"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
 
 	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
 	runtimehooksv1 "sigs.k8s.io/cluster-api/exp/runtime/hooks/api/v1alpha1"
@@ -99,7 +102,7 @@ func (e *engine) Apply(ctx context.Context, blueprint *scope.ClusterBlueprint, d
 
 		// Apply patches to the request.
 		if err := applyPatchesToRequest(ctx, req, resp); err != nil {
-			return err
+			return errors.Wrapf(err, "failed to apply patches for patch %q", clusterClassPatch.Name)
 		}
 	}
 
@@ -280,54 +283,67 @@ func createPatchGenerator(runtimeClient runtimeclient.Client, patch *clusterv1.C
 // applyPatchesToRequest updates the templates of a GeneratePatchesRequest by applying the patches
 // of a GeneratePatchesResponse.
 func applyPatchesToRequest(ctx context.Context, req *runtimehooksv1.GeneratePatchesRequest, resp *runtimehooksv1.GeneratePatchesResponse) error {
-	log := tlog.LoggerFrom(ctx)
-
 	for _, patch := range resp.Items {
-		log := log.WithValues("uid", patch.UID)
-
-		// Get the request item the patch belongs to.
-		requestItem := getRequestItemByUID(req, patch.UID)
-
-		// If a patch doesn't have a corresponding request item, the patch is invalid.
-		if requestItem == nil {
-			log.Infof("Unable to find corresponding request item with uid %q for the patch", patch.UID)
-			return errors.Errorf("unable to find corresponding request item for patch")
-		}
-
-		// Use the patch to create a patched copy of the template.
-		var patchedTemplate []byte
-		var err error
-
-		switch patch.PatchType {
-		case runtimehooksv1.JSONPatchType:
-			log.V(5).Infof("Accumulating JSON patch: %s", string(patch.Patch))
-			jsonPatch, err := jsonpatch.DecodePatch(patch.Patch)
-			if err != nil {
-				log.Infof("Failed to apply patch with uid %q: error decoding json patch (RFC6902): %s: %s", requestItem.UID, string(patch.Patch), err)
-				return errors.Wrap(err, "failed to apply patch: error decoding json patch (RFC6902)")
-			}
-
-			patchedTemplate, err = jsonPatch.Apply(requestItem.Object.Raw)
-			if err != nil {
-				log.Infof("Failed to apply patch with uid %q: error applying json patch (RFC6902): %s: %s", requestItem.UID, string(patch.Patch), err)
-				return errors.Wrap(err, "failed to apply patch: error applying json patch (RFC6902)")
-			}
-		case runtimehooksv1.JSONMergePatchType:
-			log.V(5).Infof("Accumulating JSON merge patch: %s", string(patch.Patch))
-			patchedTemplate, err = jsonpatch.MergePatch(requestItem.Object.Raw, patch.Patch)
-			if err != nil {
-				log.Infof("Failed to apply patch with uid %q: error applying json merge patch (RFC7386): %s: %s", requestItem.UID, string(patch.Patch), err)
-				return errors.Wrap(err, "failed to apply patch: error applying json merge patch (RFC7386)")
-			}
-		}
-
-		// Overwrite the spec of template.Template with the spec of the patchedTemplate,
-		// to ensure that we only pick up changes to the spec.
-		if err := patchTemplateSpec(&requestItem.Object, patchedTemplate); err != nil {
-			log.Infof("Failed to apply patch to template %s: %s", requestItem.UID, err)
-			return errors.Wrap(err, "failed to apply patch to template")
+		if err := applyPatchToRequest(ctx, req, patch); err != nil {
+			return err
 		}
 	}
+	return nil
+}
+
+func applyPatchToRequest(ctx context.Context, req *runtimehooksv1.GeneratePatchesRequest, patch runtimehooksv1.GeneratePatchesResponseItem) (reterr error) {
+	log := tlog.LoggerFrom(ctx).WithValues("uid", patch.UID)
+
+	defer func() {
+		if r := recover(); r != nil {
+			log.Infof("Observed a panic when applying patch: %v\n%s", r, string(debug.Stack()))
+			reterr = kerrors.NewAggregate([]error{reterr, fmt.Errorf("observed a panic when applying patch: %v", r)})
+		}
+	}()
+
+	// Get the request item the patch belongs to.
+	requestItem := getRequestItemByUID(req, patch.UID)
+
+	// If a patch doesn't have a corresponding request item, the patch is invalid.
+	if requestItem == nil {
+		log.Infof("Unable to find corresponding request item with uid %q for the patch", patch.UID)
+		return errors.Errorf("unable to find corresponding request item for patch")
+	}
+
+	// Use the patch to create a patched copy of the template.
+	var patchedTemplate []byte
+	var err error
+
+	switch patch.PatchType {
+	case runtimehooksv1.JSONPatchType:
+		log.V(5).Infof("Accumulating JSON patch: %s", string(patch.Patch))
+		jsonPatch, err := jsonpatch.DecodePatch(patch.Patch)
+		if err != nil {
+			log.Infof("Failed to apply patch with uid %q: error decoding json patch (RFC6902): %s: %s", requestItem.UID, string(patch.Patch), err)
+			return errors.Wrap(err, "failed to apply patch: error decoding json patch (RFC6902)")
+		}
+
+		patchedTemplate, err = jsonPatch.Apply(requestItem.Object.Raw)
+		if err != nil {
+			log.Infof("Failed to apply patch with uid %q: error applying json patch (RFC6902): %s: %s", requestItem.UID, string(patch.Patch), err)
+			return errors.Wrap(err, "failed to apply patch: error applying json patch (RFC6902)")
+		}
+	case runtimehooksv1.JSONMergePatchType:
+		log.V(5).Infof("Accumulating JSON merge patch: %s", string(patch.Patch))
+		patchedTemplate, err = jsonpatch.MergePatch(requestItem.Object.Raw, patch.Patch)
+		if err != nil {
+			log.Infof("Failed to apply patch with uid %q: error applying json merge patch (RFC7386): %s: %s", requestItem.UID, string(patch.Patch), err)
+			return errors.Wrap(err, "failed to apply patch: error applying json merge patch (RFC7386)")
+		}
+	}
+
+	// Overwrite the spec of template.Template with the spec of the patchedTemplate,
+	// to ensure that we only pick up changes to the spec.
+	if err := patchTemplateSpec(&requestItem.Object, patchedTemplate); err != nil {
+		log.Infof("Failed to apply patch to template %s: %s", requestItem.UID, err)
+		return errors.Wrap(err, "failed to apply patch to template")
+	}
+
 	return nil
 }
 

